numpyro/infer/mcmc.py

```diff
@@ -615,6 +621,260 @@ def __init__(self,
+                kwargs = {} if model_kwargs is None else model_kwargs
+                pe_fn = potential_fn_gen(*model_args, **kwargs)
+        rng_key_sa, rng_key_zs, rng_key_z = random.split(rng_key, 3)
+        z = init_params
+        z_flat, unravel_fn = ravel_pytree(z)
+        if inverse_mass_matrix is None:
+            inverse_mass_matrix = np.identity(z_flat.shape[-1]) if dense_mass else np.ones(z_flat.shape[-1])
+        inv_mass_matrix_sqrt = np.linalg.cholesky(inverse_mass_matrix) if dense_mass \
+            else np.sqrt(inverse_mass_matrix)
+        if adapt_state_size is None:
+            # XXX: heuristic choice
+            adapt_state_size = 2 * z_flat.shape[-1]
+        else:
+            assert adapt_state_size > 1, 'adapt_state_size should be greater than 1.'
+        # NB: mean is init_params
+        zs = z_flat + _sample_proposal(inv_mass_matrix_sqrt, rng_key_zs, (adapt_state_size,))
+        # compute potential energies
+        pes = lax.map(lambda z: pe_fn(unravel_fn(z)), zs)
+        if dense_mass:
+            cov = np.cov(zs, rowvar=False, bias=True)
+            if cov.shape == ():  # JAX returns scalar for 1D input
+                cov = cov.reshape((1, 1))
+            inv_mass_matrix_sqrt = np.linalg.cholesky(cov)
+        else:
+            inv_mass_matrix_sqrt = np.std(zs, 0)
+        adapt_state = SAAdaptState(zs, pes, np.mean(zs, 0), inv_mass_matrix_sqrt)
+        k = categorical_logits(rng_key_z, np.zeros(zs.shape[0]))
+        z = unravel_fn(zs[k])
+        pe = pes[k]
+        sa_state = SAState(0, z, pe, 0., 0., False, adapt_state, rng_key_sa)
+        return device_put(sa_state)
+
+    def sample_kernel(sa_state, model_args=(), model_kwargs=None):
+        pe_fn = potential_fn
+        if potential_fn_gen:
+            pe_fn = potential_fn_gen(*model_args, **model_kwargs)
+        zs, pes, loc, inv_mass_matrix_sqrt = sa_state.adapt_state
+        rng_key, rng_key_z, rng_key_reject, rng_key_accept = random.split(sa_state.rng_key, 4)
+        z_old, unravel_fn = ravel_pytree(sa_state.z)
+
+        z = loc + _sample_proposal(inv_mass_matrix_sqrt, rng_key_z)
+        pe = pe_fn(unravel_fn(z))
+        pe = np.where(np.isnan(pe), np.inf, pe)
+        diverging = (pe - sa_state.potential_energy) > max_delta_energy
+
+        locs, scales = _get_proposal_loc_and_scale(zs, loc, inv_mass_matrix_sqrt, z)
+        zs_ = np.concatenate([z[None, :], zs])
+        pes_ = np.concatenate([pe[None], pes])
+        if inv_mass_matrix_sqrt.ndim == 2:  # dense_mass
+            log_weights = dist.MultivariateNormal(locs, scale_tril=scales).log_prob(zs_) + pes_
+        else:
+            log_weights = dist.Normal(locs, scales).log_prob(zs_).sum(-1) + pes_
+        log_weights = np.where(np.isnan(log_weights), -np.inf, log_weights)
+        j = categorical_logits(rng_key_reject, log_weights)
+        # XXX: numpy.delete is not available in JAX
+        zs = np.where(np.arange(zs.shape[0])[:, None] < j, zs_[:-1], zs_[1:])
```

Comment:
should this complex logic be a helper fn?

Reply:
agreed! I'll separate it out.